components/region_cache_memory_engine/src/background.rs

```diff
@@ -451,6 +618,76 @@ impl BackgroundRunnerCore {
         #[cfg(test)]
         flush_epoch();
     }
+
+    fn evict_on_soft_limit_reached(&mut self) {
+        if self.range_stats_manager.is_none() {
+            warn!("range stats manager is not initialized, cannot evict on soft limit reached");
+            return;
+        }
+        let range_stats_manager = self.range_stats_manager.as_ref().unwrap();
+        let to_shrink_by = self
+            .memory_controller
+            .mem_usage()
+            .checked_sub(self.memory_controller.soft_limit_threshold());
+        if to_shrink_by.is_none() {
+            return;
+        }
+        let mut remaining = to_shrink_by.unwrap();
+        let mut ranges_to_evict = Vec::<(CacheRange, u64)>::with_capacity(256);
+        range_stats_manager.collect_candidates_for_eviction(&mut ranges_to_evict);
+        let ranges_to_evict = ranges_to_evict
+            .iter()
+            .filter(|(range, _)| self.engine.read().range_manager().contains_range(range))
+            .rev();
+        for (range, approx_size) in ranges_to_evict {
+            if remaining == 0 {
+                break;
+            }
+            if self.engine.write().mut_range_manager().evict_range(range) {
+                info!("evict on soft limit reached"; "range" => ?&range, "approx_size" => approx_size, "remaining" => remaining);
+                remaining = remaining
+                    .checked_sub(*approx_size as usize)
+                    .unwrap_or_default();
+                range_stats_manager.handle_range_evicted(range);
+            }
+        }
+    }
+
+    fn top_regions_load_evict(&self) {
+        if self.range_stats_manager.is_none() {
+            return;
+        }
+        let range_stats_manager = self.range_stats_manager.as_ref().unwrap();
+        if range_stats_manager.checking_top_regions() {
+            return;
+        }
+        range_stats_manager.set_checking_top_regions(true);
+
+        let curr_memory_usage = self.memory_controller.mem_usage();
+        let threshold = self.memory_controller.soft_limit_threshold();
+        range_stats_manager.adjust_num_regions(curr_memory_usage, threshold);
+
+        let mut ranges_to_add = Vec::<CacheRange>::with_capacity(256);
+        let mut ranges_to_remove = Vec::<CacheRange>::with_capacity(256);
+        range_stats_manager.collect_changed_ranges(&mut ranges_to_add, &mut ranges_to_remove);
+        info!("load_evict"; "ranges_to_add" => ?&ranges_to_add, "ranges_to_remove" => ?&ranges_to_remove);
+        for evict_range in ranges_to_remove {
```

Comment:
I suggest not to evict so eagly when the regions are not top n regions if the memory quota is still pretty sufficient. Or we may have some regions frequently load and evict again and again.

Reply:
Done.